# Importing libraries

In [2]:
#import libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [4]:
df_prods=pd.read_csv(r'D:\Sonal\career foundry\CareerFoundry\Data Immersion\Achievement 4\Exc 4.3\Original data\4.3_orders_products\products.csv',index_col=False)

In [69]:
df_ords=pd.read_csv(r'D:\Sonal\career foundry\CareerFoundry\Data Immersion\Achievement 4\Exc 4.3\Prepared data\orders_wrangled.csv',index_col=0)

In [71]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# create a dataframe

In [8]:
df_test = pd.DataFrame()

In [9]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Mixed data types

In [73]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


 Note that the original '.appymap' argument from the exercise has been deprecated and '.map' should be used instead

In [79]:
df_test['mix'] = df_test['mix'].astype('str')

 Sometimes the values will be either strings/objects or int64/floats

# Check Missing values

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape


(49693, 5)

In [18]:
# non  missing values in new df
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [19]:
df_prods_clean.shape

(49677, 5)

# Finding duplicates

In [21]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [22]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Deleting duplicates

In [24]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [25]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task

# Run the df.describe() function on your df_ords dataframe.

In [85]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Analysis and Observations:

1)'order_number'- The minimum value is 1, which is expected for the first order. However, the maximum value is 100, which seems unusually high for the number of orders placed by a single user.

2)'orders_day_of_week':(as stated in brief 0 = Saturday, 1 = sunday, etc.) 

3)'order_hour_of_day':Range: 0 to 23 (within expected range)

4)'days_since_prior_order':This column represents the number of days since the last order.
Min: 0 (valid for same-day orders)
Max: 30 (could be a data cap, warrants investigation)


# Check for mixed-type data in your df_ords dataframe

In [91]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no mix types data

In [95]:
df_ords.dtypes

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [11]:
# Remove 'eval_set' column

In [127]:
df_ords_new=df_ords.drop(columns=['eval_set'])

In [129]:
df_ords_new

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [101]:
# Change data type of 'order_id' and 'user_id'

In [131]:
df_ords_new['order_id']=df_ords_new['order_id'].astype('str')

In [133]:
df_ords_new['user_id']=df_ords_new['user_id'].astype('str')

# Run a check for missing values in your df_ords dataframe.

In [135]:
df_ords_new.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

 The 'days_since_prior_order' column is missing 206209 values

# Addressing missing values


In [137]:
df_ords_nan=df_ords_new[df_ords_new['days_since_prior_order'].isnull()== True]

In [139]:
df_ords_nan

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
11       2168274       2             1                   2                 11   
26       1374495       3             1                   1                 14   
39       3343014       4             1                   6                 11   
45       2717275       5             1                   3                 12   
...          ...     ...           ...                 ...                ...   
3420930   969311  206205             1                   4                 12   
3420934  3189322  206206             1                   3                 18   
3421002  2166133  206207             1                   6                 19   
3421019  2227043  206208             1                   1                 15   
3421069  3154581  206209             1                   3                 11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

 An explanation for this quanitty of missing values could be new customer.
 However, this is relevant information that cannot be removed.
 can be filled  the missing values with 0 (assuming that 0 represents a new customer with no prior orders.

# Run a check for duplicate values in your df_ords data.

In [148]:
df_ords_dup=df_ords_new[df_ords_new.duplicated()]

In [150]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates in df_ords.

In [156]:
df_ords_new.shape

(3421083, 6)

# Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder

In [163]:
df_ords_new.to_csv(r'D:\Sonal\career foundry\CareerFoundry\Data Immersion\Achievement 4\Exc 4.3\Prepared data\orders_cleaned.csv')

In [165]:
df_prods_clean_no_dups.to_csv(r'D:\Sonal\career foundry\CareerFoundry\Data Immersion\Achievement 4\Exc 4.3\Prepared data\products_cleaned.csv')